src/openrct2/ReplayManager.cpp

```diff
@@ -0,0 +1,775 @@
+
+        virtual bool StopPlayback() override
+        {
+            if (_mode != ReplayMode::PLAYING && _mode != ReplayMode::NORMALISATION)
+                return false;
+
+            // When normalizing the output we don't touch the mode.
+            if (_mode != ReplayMode::NORMALISATION)
+            {
+                _mode = ReplayMode::NONE;
+            }
+
+            _currentReplay.reset();
+
+            return true;
+        }
+
+        virtual bool NormaliseReplay(const std::string& file, const std::string& outFile) override
+        {
+            _mode = ReplayMode::NORMALISATION;
+
+            if (StartPlayback(file) == false)
+            {
+                return false;
+            }
+
+            if (StartRecording(outFile, k_MaxReplayTicks) == false)
+            {
+                StopPlayback();
+                return false;
+            }
+
+            _nextReplayTick = gCurrentTicks + 1;
+
+            return true;
+        }
+
+    private:
+        bool ConvertDeprecatedGameCommand(const ReplayCommand& command, ReplayCommand& result)
+        {
+            // NOTE: If game actions are being ported it is required to implement temporarily
+            //       a mapping from game command to game action. This will allow the normalisation
+            //       stage to save a new replay file with the game action being used instead of the
+            //       old game command. Once normalised the code will be no longer required.
+
+            /* Example case
+            case GAME_COMMAND_RAISE_WATER:
+            {
+                uint32_t param1 = command.ebp;
+                uint32_t param2 = command.edi;
+                result.action = std::make_unique<LandRaiseWaterAction>(param1, param2, ...);
+            }
+            */
+
+            switch (command.esi)
+            {
+                case GAME_COMMAND_COUNT: // prevent default without case warning.
+                    break;
+                default:
+                    throw std::runtime_error("Deprecated game command requires replay translation.");
+            }
+
+            return true;
+        }
+
+        bool TranslateDeprecatedGameCommands(ReplayRecordData& data)
+        {
+            for (auto&& replayCommand : data.commands)
+            {
+                if (replayCommand.action == nullptr)
+                {
+                    // Check if we can create a game action with the command id.
+                    uint32_t commandId = replayCommand.esi;
+                    if (GameActions::IsValidId(commandId))
+                    {
+                        // Convert
+                        ReplayCommand converted;
+                        if (!ConvertDeprecatedGameCommand(replayCommand, converted))
+                        {
+                            return false;
+                        }
+
+                        data.commands.erase(replayCommand);
+                        data.commands.emplace(std::move(converted));
```

Comment:
This doesn't look safe, you're modifying the container you're iterating over.

Reply:
True, never got to test this example code. It should be fine to use the iterator returned by erase I guess.

Reply:
`erase` will return the iterator to one-after-erased, so you either have to use it in a more regular `for` loop and substitute the pointer or perhaps you could simply replace the element to erase with the new one?